demo worker

UziTech opened the pull request:
## Description

I added a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Worker) to the demo page. This will allow easier testing for redos attacks.

You can test it out at: https://docs-8bxhez41v.now.sh/demo/

This PR adds is branched off #1415 with https://github.com/remyhunt/marked/pull/1 merged so we should get that PR merged first.

## Contributor

- [x] no tests required for this PR.

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

UziTech commented:
Check out the current redos [here](https://docs-9lfy8lyt6.now.sh/demo/?text=*%20%E4%BC%AA%E7%B1%BB%EF%BC%9A%5B%3Aactive%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aactive)%E3%80%81%5B%3Aany-link%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aany-link)%E3%80%81%5B%3Ablank%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ablank)%E3%80%81%5B%3Achecked%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Achecked)%E3%80%81%5B%3Acurrent%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Acurrent)%E3%80%81%5B%3Adefault%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adefault)%E3%80%81%5B%3Adefined%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adefined)%E3%80%81%5B%3Adir()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adir)%E3%80%81%5B%3Adisabled%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adisabled)%E3%80%81%5B%3Adrop%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adrop)%E3%80%81%5B%3Aempty%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aempty)%E3%80%81%5B%3Aenabled%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aenabled)%E3%80%81%5B%3Afirst%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst)%E3%80%81%5B%3Afirst-child%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst-child)%E3%80%81%5B%3Afirst-of-type%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst-of-type)%E3%80%81%5B%3Afullscreen%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afullscreen)%E3%80%81%5B%3Afuture%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afuture)%E3%80%81%5B%3Afocus%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus)%E3%80%81%5B%3Afocus-visible%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus-visible)%E3%80%81%5B%3Afocus-within%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus-within)%E3%80%81%5B%3Ahas()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahas)%E3%80%81%5B%3Ahost%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost)%E3%80%81%5B%3Ahost()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost())%E3%80%81%5B%3Ahost-context()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost-context())%E3%80%81%5B%3Ahover%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahover)%E3%80%81%5B%3Aindeterminate%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aindeterminate)%E3%80%81%5B%3Ain-range%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ain-range)%E3%80%81%5B%3Ainvalid%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ainvalid)%E3%80%81%5B%3Ais()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ais)%E3%80%81%5B%3Alang()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alang)%E3%80%81%5B%3Alast-child%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alast-child)%E3%80%81%5B%3Alast-of-type%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alast-of-type)%E3%80%81%5B%3Aleft%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aleft)%E3%80%81%5B%3Alink%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alink)%E3%80%81%5B%3Alocal-link%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alocal-link)%E3%80%81%5B%3Anot()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anot)%E3%80%81%5B%3Anth-child()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-child)%E3%80%81%5B%3Anth-col()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-col)%E3%80%81%5B%3Anth-last-child()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-child)%E3%80%81%5B%3Anth-last-col()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-col)%E3%80%81%5B%3Anth-last-of-type()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-of-type)&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22renderer%22%3A%20null%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=0.6.0)

This takes about 10 seconds to parse on my machine.

UziTech commented:
I added the ability to add commit version so we could test PRs

check out the current redos with PR #1414 [here](https://docs-id420rip6.now.sh/demo/?text=*%20%E4%BC%AA%E7%B1%BB%EF%BC%9A%5B%3Aactive%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aactive)%E3%80%81%5B%3Aany-link%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aany-link)%E3%80%81%5B%3Ablank%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ablank)%E3%80%81%5B%3Achecked%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Achecked)%E3%80%81%5B%3Acurrent%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Acurrent)%E3%80%81%5B%3Adefault%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adefault)%E3%80%81%5B%3Adefined%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adefined)%E3%80%81%5B%3Adir()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adir)%E3%80%81%5B%3Adisabled%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adisabled)%E3%80%81%5B%3Adrop%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Adrop)%E3%80%81%5B%3Aempty%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aempty)%E3%80%81%5B%3Aenabled%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aenabled)%E3%80%81%5B%3Afirst%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst)%E3%80%81%5B%3Afirst-child%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst-child)%E3%80%81%5B%3Afirst-of-type%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afirst-of-type)%E3%80%81%5B%3Afullscreen%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afullscreen)%E3%80%81%5B%3Afuture%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afuture)%E3%80%81%5B%3Afocus%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus)%E3%80%81%5B%3Afocus-visible%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus-visible)%E3%80%81%5B%3Afocus-within%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Afocus-within)%E3%80%81%5B%3Ahas()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahas)%E3%80%81%5B%3Ahost%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost)%E3%80%81%5B%3Ahost()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost())%E3%80%81%5B%3Ahost-context()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahost-context())%E3%80%81%5B%3Ahover%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ahover)%E3%80%81%5B%3Aindeterminate%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aindeterminate)%E3%80%81%5B%3Ain-range%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ain-range)%E3%80%81%5B%3Ainvalid%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ainvalid)%E3%80%81%5B%3Ais()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Ais)%E3%80%81%5B%3Alang()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alang)%E3%80%81%5B%3Alast-child%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alast-child)%E3%80%81%5B%3Alast-of-type%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alast-of-type)%E3%80%81%5B%3Aleft%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Aleft)%E3%80%81%5B%3Alink%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alink)%E3%80%81%5B%3Alocal-link%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Alocal-link)%E3%80%81%5B%3Anot()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anot)%E3%80%81%5B%3Anth-child()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-child)%E3%80%81%5B%3Anth-col()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-col)%E3%80%81%5B%3Anth-last-child()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-child)%E3%80%81%5B%3Anth-last-col()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-col)%E3%80%81%5B%3Anth-last-of-type()%5D(https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FCSS%2F%3Anth-last-of-type)&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=ceee1784fcea181890e300e9646625da9c87a3ab)

You can switch between ceee178 and master to see the redos

styfle commented on docs/demo/demo.js:
```diff
@@ -7,30 +7,44 @@ if (!window.fetch) {
 });
 
-if ('text' in search) {
+if ('text' in search && search.text) {
```
Is it necessary to do both checks here?
I would think `if (search.text)` would be sufficient.

UziTech replied:
i changed it

UziTech commented:
I also got it working with PRs

![image](https://user-images.githubusercontent.com/97994/52805239-c65a5500-304b-11e9-8e28-a746a431c6ce.png)

I moved quite a bit of code around to make it more readable and added a loading screen at the beginning since we need to load the versions and worker before anything will work.

the latest version is at: https://docs-g29qo7nxq.now.sh/demo/

UziTech commented:
If we refactor marked in multiple files and use ES modules it should still work in modern browsers.

UziTech commented:
@joshbruce correct, this only changes the demo site

joshbruce commented:
Sorry didn't notice this the first time but I'd like to add the same caveat for my approval as what's on #1414 - what's up with Snyk? What needs to happen? What can I do to help that endeavor?

UziTech commented:
see my comment on Snyk at https://github.com/markedjs/marked/pull/1414#issuecomment-464285473
